### PR TITLE
feat(spurctld): enforce real QoS limits via controller-side QosCache

### PR DESCRIPTION
**Pull request by yansun1996**
**Draft** — scoped slice of #307 (QoS group of the pending-reason emission umbrella).

> ⚠️ **Stacked on #301** (`parity/reason-code-vocab`), which is not yet merged. Until #301 lands, this PR's diff includes #301's commits. Review the QoS-emission commit (`feat(spurctld): enforce real QoS limits via controller-side QosCache`) — the rest belongs to #301. Will rebase onto `main` once #301 merges.

## Problem
#301 landed 5 QoS pending-reasons that were **wired but inert**: `qos_block_for` used a limitless `Qos::default()`, so they never fired.

## Changes
- **new `limits_cache.rs`**: `QosCache` mirroring `fairshare_cache` — `RwLock<HashMap<String, Qos>>`, `ListQos` refresh loop (timeout + stale retention), `QosInfo`→`Qos` converter.
- **cluster.rs**: `qos_cache` field + `resolve_qos()`; `qos_block_for` takes the resolved `Qos`; both call sites (`pending_jobs`, `tag_blocked_pending_reasons`) updated.
- **proto/spurdbd/sacctmgr**: extend `QosInfo` + `CreateQosRequest` with `max_submit_jobs_per_user` and `max_tres_per_user` (fields 9/10) so all five limits round-trip.
- **fix(spurdbd)**: `list_qos` decoded `usage_factor` (schema `REAL`/f32) as `f64` and **panicked**, resetting the `ListQos` h2 stream and leaving the QoS cache permanently empty. Read as `f32` and widen. **Found via live testing** — offline tests miss it because they don't cross the gRPC+Postgres boundary.

## Tests
- Unit: 5 in `limits_cache.rs` (proto→Qos, zero-means-none, cache get, end-to-end composition for submit + cpu-per-user reasons).
- Integration: `tag_blocked_sets_qos_reason_from_cache` (real `ClusterManager`, cache-seeded, asserts `QOSMaxWallDurationPerJobLimit`).
- Regression: `#[ignore]`-gated `list_qos_round_trips_all_limits` (would have caught the f32 panic).
- e2e: `test_accounting.py` QoS wall-cap reason.

## Live validation (Spur node)
- 4-cpu job in a `cpu=2` QoS → `PENDING Reason=QOSMaxCpuPerJobLimit` ✅
- 10-min job (`-t 00:10:00`) in a 1-min QoS → `PENDING Reason=QOSMaxWallDurationPerJobLimit` ✅
- Confirmed the f32 fix: before it, every `ListQos` panicked spurdbd and the cache stayed empty (jobs ran unblocked); after, QoS loads and reasons fire.

## Out of scope (boundary)
Associations (#283), partition/system limits, burst buffer (#309), unknown-QoS rejection (#282 — unknown names still get the limitless default, preserving prior behavior).

Verification: `cargo build`/`test`/`clippy --all-targets`/`fmt --check` clean; spur-core 197, spurctld 145, spurdbd 1 (+2 ignored), spur-cli 51 pass.

### CODE REVIEW

**Comment by yansun1996** [comment]
## Review addressed + live-validated (commit 189ff6d)

Independent review: **APPROVE-WITH-NITS**, no blocking issues. Nits fixed:
- Corrected misleading "~2s" QoS-cache refresh comments (the loop floors the interval at ~10s).
- Shared harness `--migrate` fix (same as #324).

### Live validation (Spur node, **#325 build** deployed fresh)
- `ListQos` works (the `usage_factor` f32 panic fix holds — `UsageFactor 1.0` renders, no h2 reset).
- 4-CPU job in QoS `cpucap` (MaxTRES cpu=2) → `PENDING Reason=QOSMaxCpuPerJobLimit` ✅
- 10-min job in QoS `short` (maxwall=1) → `PENDING Reason=QOSMaxWallDurationPerJobLimit` ✅

Validated with the #325 build specifically (separate from #324's build).

Static: build/clippy/fmt clean; spur-core 197, spurctld 145, spurdbd 1(+2 ignored), spur-cli 51 pass.

**Comment by yansun1996** [comment]
## Group 1 remainder added (commit ec91c05) — node + group QoS reasons, live-verified

Extended the QoS slice with reasons that have a real emission path:
- **`QOSMaxNodePerJobLimit`** — per-job node cap (`max_tres_per_job` Node TRES).
- **`QOSGrpCpuLimit` / `QOSGrpMemLimit` / `QOSGrpNodeLimit`** — `grp_tres` aggregated across **all running jobs in the QoS** (not just the requesting user). `grp_tres` now round-trips through `QosInfo`/`CreateQosRequest` (proto field 11), spurdbd, and `sacctmgr grptres=`.

`check_qos_limits` gained a `qos_running_tres` arg; `qos_block_for` computes both per-user and QoS-wide running TRES via a shared `sum_running_tres` helper.

### Deferred (would be dead variants — no clean data model / no distinct Slurm trigger)
`QOSGrpJobsLimit`/`QOSGrpSubmitJobsLimit` (need persisted QoS-wide job counts), `QOSGrpWallLimit` (needs consumed-wall accounting), and the coarse `QOSJobLimit`/`QOSResourceLimit`/`QOSTimeLimit` fallbacks. These land with their enforcement later.

### Tests
Unit (qos.rs): node cap + grp-cpu. Integration (spurctld): `tag_blocked_sets_qos_grp_cpu_reason` (running job fills grp cap → next job blocks). DB regression asserts `grp_tres` round-trip. Vocab/serde tables extended for the 4 variants.

### Live validation (Spur node, this build)
- 2-node job in QoS `nodecap` (node=1) → `PENDING Reason=QOSMaxNodePerJobLimit` ✅
- job in QoS `grpcap` (grp cpu=4) while a running 4-cpu job fills the group → `PENDING Reason=QOSGrpCpuLimit` ✅ (proves QoS-wide aggregation)

Static: build/clippy/fmt clean; spur-core + spurctld 199 + spurdbd + spur-cli pass.

File: crates/spurdbd/src/server.rs [hunk 1]
```diff
@@ -386,6 +386,21 @@ impl SlurmAccounting for AccountingService {
         } else {
             Some(req.max_tres_per_job.as_str())
         };
+        let max_submit = if req.max_submit_jobs_per_user == 0 {
+            None
+        } else {
+            Some(req.max_submit_jobs_per_user as i32)
+        };
```
[the remark this answers is not in the document]
Comment (yansun1996) [reply crates/spurdbd/src/server.rs:401]: Fixed in c5499f2. Replaced all three unchecked `as i32` casts in `create_qos` (`max_jobs_per_user`, `max_wall_minutes`, `max_submit_jobs_per_user`) with `i32::try_from(...)` returning an `invalid_argument` status on overflow.

File: crates/spurctld/src/cluster.rs [hunk 2]
```diff
@@ -2456,13 +2472,14 @@ fn license_block(job: &Job, pool: &HashMap<String, u64>) -> Option<spur_core::jo
...
-    // No QoS -> no QoS-based block.
-    job.spec.qos.as_ref()?;
+/// Shared by `pending_jobs()` and `tag_blocked_pending_reasons()` so the drop
+/// decision and displayed reason always agree. Caller resolves the `Qos`.
+fn qos_block_for(
+    job: &Job,
+    qos: &Qos,
+    jobs: &HashMap<JobId, Job>,
+) -> Option<spur_core::job::PendingReason> {
+    let qos_name = job.spec.qos.as_ref()?;
```
Comment (shiv-tyagi) [inline crates/spurctld/src/cluster.rs:2482]: `running_count` and `submitted_count` filter by user but not by QoS. In Slurm, `MaxJobsPerUser` and `MaxSubmitJobsPerUser` are QoS-scoped: each QoS maintains its own per-user counters (`qos_ptr->usage->user_limit_list`). A user with 2 running jobs in QoS "batch" should still be able to submit to QoS "normal" (even if "normal" also has `max_jobs=2`).

The current code aggregates across all QoS groups, making the limits overly restrictive. Pre-existing, but now that this PR activates real limits it becomes user-visible.

Fix: add `&& j.spec.qos.as_deref() == Some(qos_name.as_str())` to both the `running_count` and `submitted_count` filters below.

File: crates/spurctld/src/limits_cache.rs [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+            max_wall_minutes: 0,
+            max_tres_per_job: String::new(),
+            max_submit_jobs_per_user: 0,
+            max_tres_per_user: String::new(),
+            grp_tres: String::new(),
+        }
+    }
+
+    #[test]
+    fn test_proto_to_qos_parses_all_five_limits() {
```
Comment (shiv-tyagi) [inline crates/spurctld/src/limits_cache.rs]: Test name says "five limits" but `qos_from_proto` now parses six proto-sourced limit fields (including `grp_tres`). The `grp_tres` round-trip isn't asserted here. Consider renaming to `test_proto_to_qos_parses_all_limits` and adding:

```rust
p.grp_tres = "cpu=64".into();
// ...
assert_eq!(
    qos.limits.grp_tres.as_ref().map(|t| t.get(TresType::Cpu)),
    Some(64)
);
```

File: crates/spur-core/src/qos.rs [hunk 4]
```diff
@@ -81,6 +87,31 @@ pub fn check_qos_limits(
...
+
+        if let Some(mem) = job.spec.memory_per_node_mb {
+            let job_mem = mem * job.spec.num_nodes as u64;
+            if grp.get(TresType::Memory) > 0
+                && qos_running_tres.get(TresType::Memory) + job_mem > grp.get(TresType::Memory)
+            {
+                return QosCheckResult::Blocked(PendingReason::QosGrpMemLimit);
+            }
+        }
+    }
```
Comment (shiv-tyagi) [inline crates/spur-core/src/qos.rs:113]: The `QosGrpNodeLimit` and `QosGrpMemLimit` paths are implemented but have no behavioral tests. Only `QosGrpCpuLimit` has unit + integration coverage (`test_blocked_by_grp_cpu`, `tag_blocked_sets_qos_grp_cpu_reason`). Adding analogous tests for the node and memory group checks would catch regressions in the boundary arithmetic.

File: crates/spurctld/src/limits_cache.rs [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+}
+
+impl Default for QosCache {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Zero/empty limit fields mean "no limit" — the accounting daemon encodes absent limits this way.
+pub fn qos_from_proto(info: QosInfo) -> Qos {
```
Comment (shiv-tyagi) [inline crates/spurctld/src/limits_cache.rs]: Nit: `qos_from_proto` is `pub` but only called within this file. `limits_cache` is a private module in a binary crate, so this can just be `fn qos_from_proto` (tests in `mod tests` access it via `use super::*`).

```suggestion
fn qos_from_proto(info: QosInfo) -> Qos {
```

**Review by shiv-tyagi — COMMENTED** [review]
PTAL.

File: crates/spurctld/src/cluster.rs (hunk 2 above)
Comment (yansun1996) [reply crates/spurctld/src/cluster.rs:2482]: Fixed in a1bdcce. Added `&& j.spec.qos.as_deref() == Some(qos_name.as_str())` to the `running_count`, `submitted_count`, and `user_running_tres` filters so per-user counts are scoped to the same QoS class.

File: crates/spurctld/src/limits_cache.rs (hunk 3 above)
Comment (yansun1996) [reply crates/spurctld/src/limits_cache.rs]: Fixed in a1bdcce. Renamed to `test_proto_to_qos_parses_all_limits` and added the `grp_tres` round-trip assertion (`assert_eq!(qos.limits.grp_tres.as_ref().map(|t| t.get(TresType::Cpu)), Some(64))`).

File: crates/spur-core/src/qos.rs (hunk 4 above)
Comment (yansun1996) [reply crates/spur-core/src/qos.rs:113]: Fixed in a1bdcce. Added `test_blocked_by_grp_node` (2 nodes already running, QoS cap 4, job requests 3 → QosGrpNodeLimit) and `test_blocked_by_grp_mem` (2 GiB running, QoS cap 4 GiB, job needs 3 GiB → QosGrpMemLimit).

File: crates/spurctld/src/limits_cache.rs (hunk 5 above)
Comment (yansun1996) [reply crates/spurctld/src/limits_cache.rs]: Fixed in a1bdcce. Removed `pub` from `qos_from_proto`; tests in `mod tests` access it via `use super::*`.

**Comment by yansun1996** [comment]
## Review addressed (commit a1bdcce)

All shiv-tyagi review comments fixed:

**QoS-scoped per-user counters** (): `running_count`, `submitted_count`, and `user_running_tres` in `qos_block_for` now filter by `j.spec.qos == job.spec.qos`. Previously these aggregated across all QoS classes, making per-user limits trigger on unrelated jobs.

**`qos_from_proto` visibility** (`limits_cache.rs`): `pub` removed; it is module-private and tests reach it via `use super::*`.

**Test completeness** (`limits_cache.rs`): renamed `test_proto_to_qos_parses_all_five_limits` → `test_proto_to_qos_parses_all_limits`, added `grp_tres` round-trip assertion.

**Grp node/mem behavioral tests** (`qos.rs`): added `test_blocked_by_grp_node` and `test_blocked_by_grp_mem` covering the QosGrpNodeLimit and QosGrpMemLimit paths.

### Live validation (this build, 10.11.98.147)
- 4-CPU job in QoS `cpulimit` (maxtresperjob=cpu=2) → `PENDING QOSMaxCpuPerJobLimit` ✅
- Filling `grpcpu` QoS (grptres=cpu=4) then submitting second job → `PENDING QOSGrpCpuLimit` ✅
- 3rd submit in `submitcap` QoS (maxsubmitjobsperuser=2) → `PENDING QOSMaxSubmitJobPerUserLimit` ✅

Static: fmt/clippy clean; 152 spurctld + 1 spurdbd + spur-core tests pass.

**Review by shiv-tyagi — APPROVED** [review]
